File: android/measure/src/main/java/sh/measure/android/tracing/SpanData.kt
package sh.measure.android.tracing

import androidx.annotation.Keep

internal data class SpanData(
    val name: String,
    val traceId: String,
    val spanId: String,
    val parentId: String?,
    val sessionId: String,
    val startTime: Long,
    val endTime: Long,
    val duration: Long,
    val status: SpanStatus,
    val attributes: MutableMap<String, Any?> = mutableMapOf(),
    val userDefinedAttrs: Map<String, Any?> = emptyMap(),
    val checkpoints: MutableList<Checkpoint> = mutableListOf(),
    val hasEnded: Boolean,
    val isSampled: Boolean,
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        @Keep
        fun fromJson(json: Map<String, Any?>): SpanData {
            return SpanData(
                name = json["name"] as String,
                traceId = json["traceId"] as String,
                spanId = json["spanId"] as String,
                parentId = json["parentId"] as? String,
                sessionId = json["sessionId"] as String,
                startTime = json["startTime"] as Long,
                endTime = json["endTime"] as Long,
                duration = json["duration"] as Long,
                status = parseSpanStatus(json["status"]),
                attributes = (json["attributes"] as? MutableMap<String, Any?>) ?: mutableMapOf(),
                userDefinedAttrs = (json["userDefinedAttrs"] as? Map<String, Any?>) ?: emptyMap(),
                checkpoints = parseCheckpoints(json["checkpoints"]),
                hasEnded = json["hasEnded"] as Boolean,
                isSampled = json["isSampled"] as Boolean,
            )
        }

        private fun parseSpanStatus(statusValue: Any?): SpanStatus {
            return when (statusValue) {
                is String -> when (statusValue.lowercase()) {
                    "unset" -> SpanStatus.Unset
                    "ok" -> SpanStatus.Ok
                    "error" -> SpanStatus.Error
                    else -> SpanStatus.Unset
                }
                is Int -> when (statusValue) {
                    0 -> SpanStatus.Unset
                    1 -> SpanStatus.Ok
                    2 -> SpanStatus.Error
                    else -> SpanStatus.Unset
                }
                else -> SpanStatus.Unset
            }
        }

        private fun parseCheckpoints(checkpointsValue: Any?): MutableList<Checkpoint> {
            val result = mutableListOf<Checkpoint>()
            if (checkpointsValue is List<*>) {
                for (item in checkpointsValue) {
                    if (item is Map<*, *>) {
                        val name = item["name"] as? String ?: continue
                        val timestamp = item["timestamp"] as? Long ?: continue
                        result.add(Checkpoint(name, timestamp))
                    }
                }
            }
            return result
        }
    }
}

File: flutter/packages/measure_flutter/lib/src/tracing/span_data.dart
import 'package:measure_flutter/src/tracing/span_status.dart';

import 'checkpoint.dart';

final class SpanData {
  final String name;
  final String traceId;
  final String spanId;
  final String? parentId;
  final int startTime;
  final int endTime;
  final int duration;
  final SpanStatus status;
  final Map<String, dynamic> attributes;
  final Map<String, dynamic> userDefinedAttrs;
  final List<Checkpoint> checkpoints;
  final bool hasEnded;
  final bool isSampled;

  const SpanData({
    required this.name,
    required this.traceId,
    required this.spanId,
    this.parentId,
    required this.startTime,
    required this.endTime,
    required this.duration,
    required this.status,
    this.attributes = const {},
    this.userDefinedAttrs = const {},
    this.checkpoints = const [],
    required this.hasEnded,
    required this.isSampled,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SpanData &&
          runtimeType == other.runtimeType &&
          name == other.name &&
          traceId == other.traceId &&
          spanId == other.spanId &&
          parentId == other.parentId &&
          startTime == other.startTime &&
          endTime == other.endTime &&
          duration == other.duration &&
          status == other.status &&
          attributes == other.attributes &&
          userDefinedAttrs == other.userDefinedAttrs &&
          checkpoints == other.checkpoints &&
          hasEnded == other.hasEnded &&
          isSampled == other.isSampled;

  @override
  int get hashCode =>
      name.hashCode ^
      traceId.hashCode ^
      spanId.hashCode ^
      parentId.hashCode ^
      startTime.hashCode ^
      endTime.hashCode ^
      duration.hashCode ^
      status.hashCode ^
      attributes.hashCode ^
      userDefinedAttrs.hashCode ^
      checkpoints.hashCode ^
      hasEnded.hashCode ^
      isSampled.hashCode;

  @override
  String toString() {
    return 'SpanData{'
        'name: $name, '
        'traceId: $traceId, '
        'spanId: $spanId, '
        'parentId: $parentId, '
        'startTime: $startTime, '
        'endTime: $endTime, '
        'duration: $duration, '
        'status: $status, '
        'attributes: $attributes, '
        'userDefinedAttrs: $userDefinedAttrs, '
        'checkpoints: $checkpoints, '
        'hasEnded: $hasEnded, '
        'isSampled: $isSampled'
        '}';
  }
}
